chaindataloader/binary.go

```diff
@@ -0,0 +1,192 @@
+)
+
+// binReader is an internal struct that holds all data the binReader needs
+// to fetch headers.
+type binReader struct {
+
+	// file represents the file to be read
+	file *os.File
+
+	// tracker keeps track of the current position of the file.
+	tracker int64
+
+	// startHeight represents the height of the first header in the file.
+	startHeight int64
+
+	// endHeight represents the height of the last header in the file.
+	endHeight int64
+
+	// offset represents the distance required to read the first header from the file.
+	offset int64
+
+	// chain represents the bitcoin network the headers in the file belong to.
+	chain wire.BitcoinNet
+}
+
+// newBinaryChainDataLoader initializes a binary reader.
+func newBinaryChainDataLoader(path string) (*binReader, error) {
+	fileFlags := os.O_RDONLY
+
+	headerFile, err := os.OpenFile(path, fileFlags, 0400)
+	if err != nil {
+		return nil, fmt.Errorf("error while retrieving file: %w", err)
+	}
+
+	// Read start height of block header file. It is stored in the first
+	// 8 bytes of the header file.
+	var start int64
+	startBytes := make([]byte, startHeightSize)
+	_, err = headerFile.Read(startBytes)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining start height of file %w", err)
+	}
+	start = int64(binary.LittleEndian.Uint64(startBytes))
+
+	// Read end height of block header file. It is stored after the eighth
+	// byte of the header file.
+	var end int64
+	endBytes := make([]byte, endHeightSize)
+	_, err = headerFile.ReadAt(endBytes, startHeightSize)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining end height of file %w", err)
+	}
+	end = int64(binary.LittleEndian.Uint64(endBytes))
+
+	// Read the bitcoin network, the headers in the header file belong to.
+	// It is stored after the sixteenth byte of the file.
+	stringBytes := make([]byte, bitcoinNetSize)
+	_, err = headerFile.ReadAt(stringBytes, endHeightSize+startHeightSize)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining network of side-load source file: %w", err)
+	}
+
+	// The bitcoin network is assigned accordingly
+	chainChar := string(stringBytes)
+
+	var chain wire.BitcoinNet
+	switch {
+	case chainChar == "t":
+		chain = wire.TestNet3
+	case chainChar == "m":
+		chain = wire.MainNet
+	case chainChar == "s":
+		chain = wire.SimNet
+	case chainChar == "r":
+		chain = wire.TestNet
+	default:
+		return nil, fmt.Errorf("read unsupported character (%s) for network of side-load source file", chainChar)
+	}
+
+	return &binReader{
+		file:        headerFile,
+		startHeight: start,
+		endHeight:   end,
+		offset:      initialOffset,
+		tracker:     1,
+		chain:       chain,
+	}, nil
+}
+
+// NextHeader fetches header from the file.
+func (b *binReader) NextHeader() (*wire.BlockHeader, error) {
+	rawBlkHdr := make([]byte, headerfs.BlockHeaderSize)
+
+	if _, err := b.file.ReadAt(rawBlkHdr, b.offset+(headerfs.BlockHeaderSize*b.tracker)); err != nil {
```

Comment:
Can pull out the arithmetic here into an intermediate variable to make things a bit easier to read.